ekf2: Add support for use of multiple GPS receivers

priseborough opened the pull request:
Submitted in response to  https://github.com/PX4/Firmware/issues/9431

This PR  gives ekf2 the ability to use two GPS receivers if available to create a blended solution where the blending weights and position/height offset correction adapts to the reported accuracy. If one receiver loses lock or data, it will automatically switch across to the good receiver.

Because a long term position offset is calculated separately for each receiver, switching between either GPS1, GPS2 and the blended solution can be performed without large position and altitude jumps.

This feature relies on use of the reported accuracy metrics contained in the vehicle_gps_position.eph, epv and s_variance_m_s messages.  The EKF2_GPS_MASK parameter controls which accuracy metrics are used to set the blending ratios. By default it is zero which disables the blending. If different model receivers are used, they must both have the same update rate and both output the accuracy metrics selected by the EKF2_GPS_MASK parameter. An RTK and non RTK receiver can be used.

The EKF2_GPS_TAU parameter controls the time constant used to adjust the position offset that is applied to each GPS solution.

The blended GPS solution is published as a third GPS instance and can be found in the log as vehicle_gps_position_2

Bench test log here: https://logs.px4.io/plot_app?log=1bfcfab1-1652-48e3-b349-a42cd493e3cd

Method used to log blended GPS to a third instance may not be correct. I am relying on the uORB orb_publish_auto function to correctly assign a new instance.

Noise on blended lat/lon in log needs investigating. It was not present on logs generated using earlier replay data.

priseborough commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -827,38 +896,114 @@ void Ekf2::run()
+				_gps_select_index = 1;
+
+			} else if (_gps_select_index == 2) {
+				// use last recever we received data from
```
will fix

priseborough commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -572,6 +628,13 @@ void Ekf2::run()
 {
 	bool imu_bias_reset_request = false;
 
+	// becasue we can have a second GPS used as a truth reference to asisst with tuning and algorithm development using replay
```
will fix

priseborough commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -1523,6 +1665,441 @@ const Vector3f Ekf2::get_vel_body_wind()
+
+		for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+			if (_gps_state[i].fix_type >= 3 && _gps_state[i].sacc >= 0.001f) {
+				spd_blend_weights[i] = speed_accuracy_sum_sq / (_gps_state[i].sacc * _gps_state[i].sacc);
```
yes that is correct

priseborough replied:
fixed

priseborough commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -1523,6 +1665,441 @@ const Vector3f Ekf2::get_vel_body_wind()
+		// report the best valid accuracies and DOP metrics
+
+		if (_gps_state[i].eph > 0.0f
+		    && _gps_state[i].eph < _gps_state[GPS_BLENDED_INSTANCE].eph) {
```
We need to take the smallest non zero value so that won't work either. I need to initialise _gps_state[GPS_BLENDED_INSTANCE].eph to the largest float.

priseborough replied:
fixed

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -572,6 +628,13 @@ void Ekf2::run()
 	bool imu_bias_reset_request = false;
 
+	// support for dual GPS inputs if available
+	int gps_subs[2];
```
delete? _gps_subs in class

priseborough replied:
I don't understand

dagar replied:
You have both gps_subs[2] and _gps_subs[ORB_MULTI_MAX_INSTANCES].

priseborough replied:
I thought I needed a separate one for  publishing. will fix.

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -572,6 +628,13 @@ void Ekf2::run()
+	int gps_subs[2];
+
+	for (unsigned i = 0; i < 2; i++) {
+		gps_subs[i] = orb_subscribe_multi(ORB_ID(vehicle_gps_position), i);
```
Should be with _gps_subs in constructor.

priseborough replied:
will fix

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -512,6 +565,8 @@ Ekf2::Ekf2():
 
 	for (unsigned i = 0; i < ORB_MULTI_MAX_INSTANCES; i++) {
+		_gps_subs[i] = -1;
+		_gps_subs[i] = orb_subscribe_multi(ORB_ID(vehicle_gps_position), i);
```
GPS_MAX_RECEIVERS for _gps_subs, not ORB_MULTI_MAX_INSTANCES.

priseborough replied:
fixed

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -584,6 +647,12 @@ void Ekf2::run()
 	sensor_selection_s sensor_selection = {};
 
+	// Position of local NED origin in GPS / WGS84 frame
+	map_projection_reference_s ekf_origin = {};
```
Move back?

priseborough replied:
I don't understand

priseborough replied:
@dagar Can you elaborate on this?

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -827,38 +896,114 @@ void Ekf2::run()
 
+		// check for second GPS receiver data
+		bool gps2_updated = false;
+		orb_check(gps_subs[1], &gps2_updated);
```
gps_subs[1] -> _gps_subs[1]

priseborough replied:
will fix

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -165,11 +198,21 @@ class Ekf2 final : public ModuleBase<Ekf2>, public ModuleParams
 	const float _hgt_innov_spike_lim = 2.0f * _hgt_innov_test_lim;	///< preflight position innovation spike limit (m)
 
+	// GPS blending and switching
+	struct gps_message _gps_state[GPS_MAX_RECEIVERS + 1]; ///< internal state data for the physical and blended GPS
```
Suggestion, the code below might be a bit easier to follow if you kept the blended GPS in a separate struct rather than the end.
`_gps_state[GPS_BLENDED_INSTANCE] -> _gps_blended`

priseborough replied:
Agree - wil fix

priseborough replied:
fixed

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -165,11 +198,21 @@ class Ekf2 final : public ModuleBase<Ekf2>, public ModuleParams
 
+	// GPS blending and switching
+	struct gps_message _gps_state[GPS_MAX_RECEIVERS + 1]; ///< internal state data for the physical and blended GPS
+	struct gps_message _gps_output[GPS_MAX_RECEIVERS + 1]; ///< output state data for the physical and blended GPS
```
Why is it necessary to store _gps_output? Couldn't it be discarded at the end of each update_gps_blend_states?

priseborough replied:
We can't discard it until after GPS data has been written to the EKF buffer.

priseborough replied:
I will refactor the code to remove the need for that class variable. Refactoring needs to happen anyway because there is a corner case that is not being handled correctly at the moment that results in the output not updating if one of the receivers loses lock. This was picked up in dynamic testing.

priseborough replied:
I've split the calc_gps_blend_outputs into 3 functions for updating of receiver offsets, application of receiver offsets and calculation of blended  receiver output.

I'm using a class variable for outputs because I am unsure how to pass an array of structs by reference without creating a new type, and the compiler would not let me #include <array> and use std::array to create a new type

RomanBapst commented:
@priseborough I'm done with review (can't approve on my phone). We will test it tomorrow.

priseborough commented:
There was a bug found during dynamic ground testing that caused the GPS output state to not be updated f one of the receivers was disconnected or lost lock. I've split the calc_gps_blend_outputs() function into three separate functions for updating steady state (filtered) offsets, correcting physical receiver data using the offsets and calculating a blended output using the corrected receiver data.

Testing on replay shows the algorithm is working properly. However, although printing the lat and lon values in the gps struct one line above where they are published shows the correct values, when I review the log generated by the ekf2 replay, the vehicle_gps_position_2.lat and lon values are not correct so I have a mistake in the publishing of the blended GPS.

priseborough commented:
The CI is showing px4fmu v2 to be out of flash space.

priseborough commented:
Bench testing at commit https://github.com/PX4/Firmware/pull/9765/commits/935b701d723720703d8f9acc5fe333d0767dba23 shows logging of the third (blended) GPS instance working as expected, so the logging problem appears to a replay issue.

The remaining issue is the large initialisation error in the blended solution for lat and lon that takes several seconds to remove - see bench test result below:

![screen shot 2018-06-28 at 3 58 45 pm](https://user-images.githubusercontent.com/3596952/42015808-2e4b2934-7aec-11e8-9c92-4ea8b9e2722b.png)

Replaying the same data gives a much smaller startup transient - see

![screen shot 2018-06-28 at 4 05 42 pm](https://user-images.githubusercontent.com/3596952/42016018-280a39ec-7aed-11e8-8a64-41eb96ff636a.png)

priseborough commented:
Last commit https://github.com/PX4/Firmware/pull/9765/commits/79f41ce25726fa9534f49ef1695cab834efeee15 seems to have resolved remaining issues. Here are GPS alt, lat and lon from the last test:

<img width="832" alt="screen shot 2018-06-28 at 8 10 56 pm" src="https://user-images.githubusercontent.com/3596952/42028478-b7d865a4-7b0f-11e8-8bd2-736ad8e65671.png">

<img width="832" alt="screen shot 2018-06-28 at 8 11 11 pm" src="https://user-images.githubusercontent.com/3596952/42028480-b99bbc7e-7b0f-11e8-93ed-92e6b3287b7f.png">

<img width="828" alt="screen shot 2018-06-28 at 8 11 24 pm" src="https://user-images.githubusercontent.com/3596952/42028487-bc638388-7b0f-11e8-993f-e0ce2845f1b0.png">

priseborough commented:
After experimenting with some logging options with bench testing, it appears that unless estimator replay data logging is enabled via SDLOG_PROFILE, then the GPS blending does not work correctly and the blended instance logged closely tracks the second receiver . I suspect that publishing of the blended instance could be overwriting data from a physical receiver.

RomanBapst commented:
@priseborough We did two flights with dual gps:

In the first one something went wrong, one gps position was showing showing somewhere in the pacific
https://review.px4.io/plot_app?log=0a88185a-28b7-4ce8-b048-328b961d44a6

The second try was good, we managed to fly in position control mode. Maybe it was drifting a bit more than usual but it might not be related to the changes
https://review.px4.io/plot_app?log=4d30ebe3-48f0-433b-8fb9-b1eaf6da5392

I haven't had a chance to look into them.

priseborough commented:
@hamishwillee Good reminder. We will be setting the parameters so that the feature is disabled by default. Is there a way to flag a feature in master as 'beta'  until we are happy it has had enough community test coverage?

hamishwillee commented:
> Good reminder. We will be setting the parameters so that the feature is disabled by default. Is there a way to flag a feature in master as 'beta' until we are happy it has had enough community test coverage?

There is no "formal" way. What I normally do is either
a) Document after it is "ready" 
b) Put a big warning near where the feature is documented. 

Is there any specialized QGC UI for setup, or is this just going to be a parameter? Obviously we now have to explain what ports you can add the second GPS to, how to turn it on, any tuning, and how to turn off anything else that might be using the port. 

I would probably add doc here: https://docs.px4.io/en/gps_compass/, and possibly also here: https://docs.px4.io/en/advanced_features/

priseborough commented:
@hamish Just one parameter for most applications. There is second time constant parameter controlling how fast the individual GPS offsets are corrected, but I am hoping we can tune that so that 99% of users don't need to touch it.

priseborough commented:
@RomanBapst The first log appears to be afflicted by the same initialisation bug affecting the lat and lon output that I was chasing and fixed yesterday. The second log shows the blending working as expected for alt, lat, lon and velocity other than 0.4m 'switching' noise on the lat/lon output that appears at certain times. Running the data through the latest version reproduces the noise so this can be investigated and fixed. It does not affect the altitude, so is not an issue with the weighting calculation and appears to be a problem in the way global coordinates are converted to local position deltas and vice versa. 

Data from the the second flight:

![screen shot 2018-06-29 at 5 33 48 pm](https://user-images.githubusercontent.com/3596952/42079289-9b4d3e74-7bc2-11e8-9041-6008f541d1b0.png)

![screen shot 2018-06-29 at 5 35 37 pm](https://user-images.githubusercontent.com/3596952/42079356-dc9198d0-7bc2-11e8-8b7c-c47a885fc520.png)

![screen shot 2018-06-29 at 5 36 52 pm](https://user-images.githubusercontent.com/3596952/42079456-1f47f0c0-7bc3-11e8-8dec-f45637ff36e4.png)

hamishwillee commented:
> @hamish Just one parameter for most applications. There is second time constant parameter controlling how fast the individual GPS offsets are corrected, but I am hoping we can tune that so that 99% of users don't need to touch it.

Thanks. YOu could put in a Note in the param definition itself warning that the feature is still not robust, but you would have to remove it later.

I suspect users will also need to use a param to turn off whatever is using the port to which you attach the second GPS. ie SYS_COMPANION

priseborough commented:
Writing the blended GPS to a third instance of the GPS message is causing problems on replay with the lat,lon values in vehicle_gps_position_2 not matching the values logged using a custom message because the replay and ekf2 are both publishing to that instance.

@dagar is there any way I can make ekf2 write the blended output to a specified instance (eg index 3) when doing replay?

priseborough commented:
@RomanBapst I have been able to replay your log successfully after modifying the ekf2 to publish the blended GPS solution to a separate topic - see https://github.com/PX4/Firmware/pull/9765/commits/3198edfd0ad6ed7e5396e99075b9f5c406c11f2b.

Here are the blended lat, lon and alt:

<img width="957" alt="screen shot 2018-06-30 at 2 25 23 pm" src="https://user-images.githubusercontent.com/3596952/42121424-1b011af0-7c72-11e8-988a-5fcd06f8cf10.png">

<img width="958" alt="screen shot 2018-06-30 at 2 25 37 pm" src="https://user-images.githubusercontent.com/3596952/42121425-1e3a559c-7c72-11e8-9b3d-a34d22ebc396.png">

<img width="951" alt="screen shot 2018-06-30 at 2 27 19 pm" src="https://user-images.githubusercontent.com/3596952/42121426-21103d0e-7c72-11e8-96f1-e79ca0c9f065.png">

priseborough commented:
@dagar and @RomanBapst The px4fmu v2 build is failing (again) due to lack of flash. I need some advice on what  we could remove from that build to get under the flash size  limit.

dagar commented:
This is a reasonable option that won't further cripple px4fmu-v2. https://github.com/PX4/Firmware/pull/9770
I'll see if we can move it forward.

dagar commented:
Would generating another version of vehicle_gps_position (eg vehicle_gps_position_blended) be sufficient instead of a new nearly duplicate message?

See vehicle_local_position as an example. https://github.com/PX4/Firmware/blob/master/msg/vehicle_global_position.msg#L30

priseborough commented:
@dagar Yes I thought of that, but don't need some of the other data that is in the vehicle_gps_position message would need to add the GPS select index.

priseborough commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -1523,6 +1665,441 @@ const Vector3f Ekf2::get_vel_body_wind()
+		}
+
+		if (_gps_state[i].epv > 0.0f
+		    && _gps_state[i].epv < _gps_state[GPS_BLENDED_INSTANCE].epv) {
```
fixed

priseborough commented:
@RomanBapst the review tracker is listing you as having outstanding change requests. Can you let me know if there are any outstanding items and if not can you please complete the approval process. TXS.

mhkabir commented:
This is just a nitpick, but what do you guys think about having the current vehicle_gps_position topic renamed to sensor_gps, in line with what we use for the raw sensor data for other sensors, and use vehicle_gps_position for the blended solution? I generally don't like the specific `ekf_* topic naming.

dagar commented:
@mhkabir I had the same thought, but I think it'll be relatively easy to move the GPS blending functionality once we make the necessary architectural changes. In the meantime this unblocks real world testing.

What we really need to figure out is a good general architecture for these sensor specific configurations. We don't want EKF2 to own the GPS positions, or waste publishing static data in every single message. I think something like a status/info message to go along with each sensor instance that only publishes on change.

dagar commented:
Does this need an additional ekf2_timestamps.gps_timestamp_rel for replay?

priseborough commented:
I modified the code to log the relative timestamps for each GPS receiver and plotted them from a bench test with a UBlox 8 connected to GPS1 and a UBlox 6 connected to GPS2 - see plot below with relative timestamp rescaled to units of msec (message is in 0.1 msec units)

![screen shot 2018-07-05 at 10 46 12 am](https://user-images.githubusercontent.com/3596952/42297164-a905b10e-8040-11e8-8a61-2efd3316bc70.png)

Not having separate replay timestamps is going to make some difference.

thomasgubler commented:
@bkueng can you look into @dagar question please?

thomasgubler commented:
TODO: @priseborough does regression test with single gps setup

priseborough commented:
Jenkins test this please

bkueng commented:
> Does this need an additional ekf2_timestamps.gps_timestamp_rel for replay?

If we want perfectly reproducible replay then yes. The difference will be small since it's a low-rate topic. What can happen without it is that IMU messages right around the gps topic publication can be fused before or after the gps message.

priseborough commented:
I will rebase and resolve conflicts arising from https://github.com/PX4/Firmware/pull/9832.

We can address the replay issue in a separate PR.

priseborough commented:
SITL log from successful reversion test here: https://logs.px4.io/plot_app?log=ace38f4b-1a75-4f03-8b97-1ba1878ad51b

Antiheavy commented:
Does this PR support a 2nd GPS receiver data over MAVlink (i.e. some generic GPS messages fed to the FMU over a MAVlink serial connection)?  If not, can it be extended to support this easily, either now or in the future?

priseborough commented:
@Antiheavy Provided the mavlink receiver published the data using the vehicle_gps uORB message to a second instance, then in principle it should still work. It would need to have the same definitions for the accuracy metrics being selected to calculate the weights , eg eph, s_variance_m_s, etc. It would also need to have the same update rate.

priseborough commented:
@RomanBapst Can you please change your review status if you are happy with the changes made. TXS.

RomanBapst commented:
@priseborough I'll review your last changes.

dagar commented:
I'll rebase to resolve the conflict.

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -847,35 +913,135 @@ void Ekf2::run()
+				}
+
+				if (_gps_blend_mask.get() == 0) {
+					// When GPS blending is disabled we always use the first receiver instance
```
Duplicate setGpsData. I'll push a fix.

dagar commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -847,35 +913,135 @@ void Ekf2::run()
+			_ekf.setGpsData(_gps_output[_gps_select_index].time_usec, &_gps_output[_gps_select_index]);
+
+			// log blended solution as a third GPS instance
+			if (_gps_blend_mask.get() > 0) {
```
Unnecessary check.

dagar commented:
Here we go again...

![image](https://user-images.githubusercontent.com/84712/42740220-12b36f04-8872-11e8-833b-af908c1d82a4.png)

I'll see what I can find.

priseborough commented:
Rebased

dagar commented:
Rebased on master and currently fitting into px4fmu-v2.

dagar commented:
Let's get this in!

lingbuweibu commented:
If the eph or epv or evh is not change,but the lon or lat is jump,how can I do?

lingbuweibu commented:
// Use the oldest non-zero time, but if time difference is excessive, use newest to prevent a disconnected receiver from blocking updates
	uint64_t max_us = 0; // newest non-zero system time of arrival of a GPS message
	uint64_t min_us = -1; // oldest non-zero system time of arrival of a GPS message

	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
		// Find largest and smallest times
		if (_gps_state[i].time_usec > max_us) {
			max_us = _gps_state[i].time_usec;
		}

		if ((_gps_state[i].time_usec < min_us) && (_gps_state[i].time_usec > 0)) {
			min_us = _gps_state[i].time_usec;
		}
	}
The min_us in here is a bug？？？

dagar commented:
@lingbuweibu no, that's initializes min_us to the largest uint64 value so that the comparison sets min_us to an actual gps time.

lingbuweibu commented:
> @lingbuweibu no, that's initializes min_us to the largest uint64 value so that the comparison sets min_us to an actual gps time.

OK
